detectors/aws/lambda/README.md

```diff
@@ -0,0 +1,67 @@
+# OpenTelemetry AWS Lambda Resource Detector for Golang
+
+[![Go Reference][goref-image]][goref-url]
+[![Apache License][license-image]][license-url]
+
+This module detects resource attributes available in AWS Lambda.
+
+## Installation
+
+```bash
+go get -u go.opentelemetry.io/contrib/detectors/aws/lambda
+```
+
+## Usage
+
+Create a sample Lambda Go application such as below.
+
+```go
+package main
+
+import (
+	"github.com/aws/aws-lambda-go/lambda"
+	sdktrace "go.opencensus.io/otel/sdk/trace"
+	lambdadetector "go.opentelemetry.io/contrib/detectors/aws/lambda"
+)
+
+func main() {
+	detector := lambdadetector.NewResourceDetector()
+	res, err := detector.Detect(context.Background())
+	if err != nil {
+		fmt.Printf("failed to detect lambda resources: %v\n", err)
+	}
+
+	tp := sdktrace.NewTracerProvider(
+		sdktrace.WithResource(res),
+	)
+	lambda.Start(<some lambda handler>)
+}
+```
+
+Now your `TracerProvider` will have the following resource attributes and attach them to new spans:
+
+| Resource Attribute | Example Value |
+| --- | --- |
+| `cloud.provider` | aws
+|`cloud.region` | us-east-1 
+|`faas.name` | MyLambdaFunction 
+|`faas.version` | $LATEST
+
+Of note, `faas.id` and `cloud.account.id` are not set by the Lambda resource detector because they are not available outside a Lambda invocation. For this reason, when using the AWS Lambda Instrumentation these attributes are set as additional span attributes.
```

Comment:
Could this be solved through lazy initialization?

Reply:
Not without asking the user to add more code. The `faas.id` and `cloud.account.id` attributes are only available inside of the handler, and our solution is meant to be strictly a wrapper of the handler such that we can't actually do anything inside the wrapper. So we could ask the user to lazy-initialize the tracer provider in their handler, but that's a much worse experience.

Reply:
Why would the *user* have to do that lazy-initialization, rather than the framework? E.g., at [this point](https://github.com/open-telemetry/opentelemetry-go-contrib/pull/983/files#diff-76cf3df45b0cb4100c653ba5800e2123291f47369b5c96e68a6df167c2403c9dR122) it seems like you'd have access to those invocation-scoped variables. It's true that you'd have to do separately for each possible reflective type of the handler function.

Reply:
I think the problem is that we will then couple the Lambda resource detector and the Lambda instrumentation, which ought to be two separate components. The Lambda instrumentation does indeed have access to those invocation-scoped attributes, and is able to initialize the tracer provider. But the Lambda resource detector must operate only with what is available in the static environment, so if a user wants to initialize a custom tracer provider outside their handler as seen in the example it will work the same.

I agree that it's unfortunate the lambda resource detector doesn't capture these attributes, but I believe this is a limitation of Lambda itself more than any individual instrumentation, which is why the span attribute is called out [in the spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/faas.md) and there's talks of [not even marking faas.id as a resource attribute](https://github.com/open-telemetry/opentelemetry-specification/issues/1261) at all.

Reply:
Thanks for explaining. I was commenting based on "what would an operator find most useful" vs "what is a clean way to factor the code." Given the ambiguity / ongoing discussion around faas.id, I think this code makes a reasonable choice.